controllers/tenant_controller.go

```diff
@@ -0,0 +1,528 @@
+	logger := log.FromContext(ctx)
+
+	tenant := &cattagev1beta1.Tenant{}
+	if err := r.client.Get(ctx, req.NamespacedName, tenant); err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	if tenant.DeletionTimestamp != nil {
+		if err := r.finalize(ctx, tenant); err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to finalize: %w", err)
+		}
+		return ctrl.Result{}, nil
+	}
+
+	defer func(before cattagev1beta1.TenantStatus) {
+		if !equality.Semantic.DeepEqual(tenant.Status, before) {
+			logger.Info("update status", "status", tenant.Status, "before", before)
+			if err2 := r.client.Status().Update(ctx, tenant); err2 != nil {
+				logger.Error(err2, "failed to update status")
+				err = err2
+			}
+		}
+	}(tenant.Status)
+
+	err = r.reconcileNamespaces(ctx, tenant)
+	if err != nil {
+		tenant.Status.Health = cattagev1beta1.TenantUnhealthy
+		meta.SetStatusCondition(&tenant.Status.Conditions, metav1.Condition{
+			Type:    cattagev1beta1.ConditionReady,
+			Status:  metav1.ConditionFalse,
+			Reason:  "Failed",
+			Message: err.Error(),
+		})
+		return ctrl.Result{}, err
+	}
+
+	err = r.reconcileArgoCD(ctx, tenant)
+	if err != nil {
+		tenant.Status.Health = cattagev1beta1.TenantUnhealthy
+		meta.SetStatusCondition(&tenant.Status.Conditions, metav1.Condition{
+			Type:    cattagev1beta1.ConditionReady,
+			Status:  metav1.ConditionFalse,
+			Reason:  "Failed",
+			Message: err.Error(),
+		})
+		return ctrl.Result{}, err
+	}
+
+	tenant.Status.Health = cattagev1beta1.TenantHealthy
+	meta.SetStatusCondition(&tenant.Status.Conditions, metav1.Condition{
+		Type:   cattagev1beta1.ConditionReady,
+		Status: metav1.ConditionTrue,
+		Reason: "OK",
+	})
+	logger.Info("Tenant successfully reconciled")
+
+	return ctrl.Result{}, nil
+}
+
+func containNamespace(roots []cattagev1beta1.RootNamespaceSpec, ns corev1.Namespace) bool {
+	for _, root := range roots {
+		if root.Name == ns.Name {
+			return true
+		}
+	}
+	return false
+}
+
+func (r *TenantReconciler) disownNamespace(ctx context.Context, ns *corev1.Namespace) error {
+	managed, err := accorev1.ExtractNamespace(ns, constants.FieldManager)
+	if err != nil {
+		return err
+	}
+	delete(managed.Labels, constants.OwnerTenant)
+	for k := range r.config.Namespace.CommonLabels {
+		delete(managed.Labels, k)
+	}
+	for k := range r.config.Namespace.CommonAnnotations {
+		delete(managed.Annotations, k)
+	}
+	err = r.patchNamespace(ctx, managed)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (r *TenantReconciler) removeRBAC(ctx context.Context, tenant *cattagev1beta1.Tenant, ns *corev1.Namespace) error {
+	rb := &rbacv1.RoleBinding{}
+	err := r.client.Get(ctx, client.ObjectKey{Namespace: ns.Name, Name: tenant.Name + "-admin"}, rb)
+	if apierrors.IsNotFound(err) {
+		return nil
+	}
+	if rb.DeletionTimestamp != nil {
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+	err = r.client.Delete(ctx, rb)
```

Comment:
Is it OK not to check the RoleBinding is managed by cattage?
Since we can manually add `cattage.cybozu.io/tenant` to any namespace, that cause the RoleBinding of the name `xxx-admin` to be removed after a resync period.